File: LeetCode-py/94_inorderTraversal.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def createTree(nodeList):
    if (not nodeList[0]):
        return None
    root = TreeNode(nodeList[0])
    i = 1
    Nodes = [root]
    for node in Nodes:
        if node != None:
            node.left = TreeNode(nodeList[i]) if nodeList[i] != None else None
            Nodes.append(node.left)
            i += 1
            if i == len(nodeList): return root
            node.right = TreeNode(nodeList[i]) if nodeList[i] != None else None
            Nodes.append((node.right))
            i += 1
            if i == len(nodeList): return root

class Solution:
    @classmethod
    def inorderTraversal(self, root: TreeNode):
        def dfs(root):
            if root!=None:
                dfs(root.left)
                res.append(root.val)
                dfs(root.right) 
        res = []
        dfs(root)
        return res

if __name__ == "__main__":
    treelist = [1, None, 2, 3]
    root = createTree(treelist)
    print(Solution.inorderTraversal(root))

File: LeetCode-c++/94_inorderTraversal.cpp
#include<iostream>
#include<vector>
#include<queue>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x): val(x), left(NULL), right(NULL){}
};

TreeNode* createTree(vector<int>& nodeList)
{
    if(nodeList[0]==NULL)
    {   
        return NULL;
    }
    TreeNode* root = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    root->val = nodeList[0];
    int i = 1;
    queue<TreeNode*> que;
    que.push(root);
    int len = nodeList.size();
    while(!que.empty())
    {
        
        TreeNode *node = que.front();
        que.pop();
        if(node != NULL)
            {
                if(i==len)
                {
                    node->left = NULL;
                    node->right = NULL;
                    continue;
                }
                if(nodeList[i])
                {
                    TreeNode* n = (struct TreeNode *)malloc(sizeof(struct TreeNode));
                    n->val = nodeList[i];
                    node->left = n;
                    que.push(node->left);
                    
                }else{
                    node->left = NULL;
                }
                i++;
                if(i==len) 
                {
                    node->right = NULL;
                    continue;
                }
                if(nodeList[i])
                {
                    TreeNode* n = (struct TreeNode *)malloc(sizeof(struct TreeNode));
                    n->val = nodeList[i];
                    node->right = n;
                    que.push(node->right);
                }else{
                    node->right = NULL;
                }
                i++;
                if(i==len) continue;
            }
            
    }
    return root;
}

class Solution
{
    public:
    vector<int> inorderTraversal(TreeNode* root)
    {
        vector<int> res;
        dfs(root, res);
        return res;
    }
    void dfs(TreeNode* root, vector<int>& res)
    {
        if(!root)
            return;
        dfs(root->left, res);
        res.push_back(root->val);
        dfs(root->right, res);
    }
    
};

int main()
{
    vector<int> treelist = {1, NULL, 2, 3};
    TreeNode* root = createTree(treelist);
    Solution solution = Solution();
    vector<int> res = solution.inorderTraversal(root);

    cout << '[';
    for(int x:res)
    {
        cout << x <<' ';
    }
    cout << ']' << endl;
    system("pause");
    return 0;
}